core/src/main/scala/org/apache/spark/rpc/netty/Dispatcher.scala

```diff
@@ -0,0 +1,226 @@
+      }
+      endpointToEndpointRef.put(endpoint, endpointRef)
+      val inbox = new Inbox(endpointRef, endpoint)
+      endpointToInbox.put(endpoint, inbox)
+      receivers.put(inbox.endpoint)
+    }
+    endpointRef
+  }
+
+  def getRpcEndpointRef(endpoint: RpcEndpoint): RpcEndpointRef = endpointToEndpointRef.get(endpoint)
+
+  def removeRpcEndpointRef(endpoint: RpcEndpoint): Unit = endpointToEndpointRef.remove(endpoint)
+
+  def getRpcEndpointRef(name: String): RpcEndpointRef = nameToEndpoint.get(name).endpointRef
+
+  // Should be idempotent
+  private def unregisterRpcEndpoint(name: String): Unit = {
+    val endpointPair = nameToEndpoint.remove(name)
+    if (endpointPair != null) {
+      val inbox = endpointToInbox.get(endpointPair.endpoint)
+      if (inbox != null) {
+        inbox.stop()
+        receivers.put(inbox.endpoint)
+      }
+    }
+  }
+
+  def stop(rpcEndpointRef: RpcEndpointRef): Unit = {
+    synchronized {
+      if (stopped) {
+        // This endpoint will be stopped by Distpatcher.stop() method.
+        return
+      }
+      unregisterRpcEndpoint(rpcEndpointRef.name)
+    }
+  }
+
+  /**
+   * Send a message to all registered [[RpcEndpoint]]s.
+   * @param message
+   */
+  def broadcastMessage(message: BroadcastMessage): Unit = {
+    val iter = endpointToInbox.values().iterator()
+    while (iter.hasNext) {
+      val inbox = iter.next()
+      postMessageToInbox(inbox, message)
+    }
+  }
+
+  def postMessage(message: RequestMessage, callback: RpcResponseCallback): Unit = {
+    val receiver = nameToEndpoint.get(message.receiver.name)
+    if (receiver != null) {
+      val inbox = endpointToInbox.get(receiver.endpoint)
+      if (inbox != null) {
+        val rpcCallContext =
+          new RemoteNettyRpcCallContext(
+            nettyEnv, inbox.endpointRef, callback, message.senderAddress, message.needReply)
+        postMessageToInbox(inbox,
+          ContentMessage(message.senderAddress, message.content, message.needReply, rpcCallContext))
+        return
+      }
+    }
+    callback.onFailure(
+      new SparkException(s"Could not find ${message.receiver.name} or it has been stopped"))
+  }
+
+  def postMessage(message: RequestMessage, p: Promise[Any]): Unit = {
+    val receiver = nameToEndpoint.get(message.receiver.name)
+    if (receiver != null) {
+      val inbox = endpointToInbox.get(receiver.endpoint)
+      if (inbox != null) {
+        val rpcCallContext =
+          new LocalNettyRpcCallContext(
+            inbox.endpointRef, message.senderAddress, message.needReply, p)
+        postMessageToInbox(inbox,
+          ContentMessage(message.senderAddress, message.content, message.needReply, rpcCallContext))
+        return
+      }
+    }
+    p.tryFailure(
+      new SparkException(s"Could not find ${message.receiver.name} or it has been stopped"))
+  }
+
+  private def postMessageToInbox(inbox: Inbox, message: InboxMessage): Unit = synchronized {
```

Comment:
Shouldn't the `postMessage` methods be synchronized instead? The way it is, if you call `postMessage` on a stopped dispatcher, you won't be notified of that fact - the message will just be dropped.

There's also another race between `unregisterRpcEndpoint` can `postMessage`; you may end up posting a message to an endpoint that has been unregistered. I'm not sure whether that can cause problems (need to read the code again), but it's always fishy when things like that can happen.